### ISSUE TITLE
How do I prevent leave route?

### ISSUE THREAD

**Issue by mongris** [frontarm/navi#93]
Thanks for the nimble navigation project. I am considering use navi instead of react-router in next project.

I want to redirect new user with initial password to `change_password_page` and prevent it leave the page before password changed. I can't find the impl in navi, can you help to figure out how make it works with navi?

**Comment by jamesknelson** [comment]
One possibility would be to put the user details in context, and then use a top-level `map`. Something like this:

```js
mount({
  '/change-password': lazy(() => import('./change-password')),
  '*': map(request => {
    if (request.context.requiresPasswordChange) {
      return redirect('/change-password?redirectTo='+encodeURIComponent(request.url.href), { exact: false })
    }
    else {
      return routes
    }
  })
})
```

**Comment by mongris** [comment]
@jamesknelson Thanks for your reply.

I need custom dialog to prompt user to change password first when he click other link to redirecting.

I dig into source code then I think maybe there should add some methods in Navigation class:

unsubscriber = navigation.beforeLeave(callback return bool)



then use it like:

```javascript
useEffect(() => {
  const check = () => {
    if (context.user.requiredPasswordChanged) {
      customDialog.show()
      return false
    }
    return true
  }
  return navigation.beforeLeave(check).unsubscribe
}, [])
```

```javascript
navigation.prototype.beforeLeave = (callback) => {
  this._beforeLeaveCheck = callback
  const allowed = this._beforeLeaveCheck()
  if (!allowed) this._history.block()
  else this._history.unblock()
  return unsubscriber(this._beforeLeaveCheck)
}
```

**Comment by roborich** [comment]
I need the same functionality. I need to prompt a user to save before navigating away.  It needs to happen before the component unmounts.

**Comment by roborich** [comment]
Nevermind.  Redirecting back to the original component works.
I had something in the component watching the route that caused it to re-render.

**Comment by JacksonGariety** [comment]
Can we add a `navigation.pause()` method that sets the context within navi?

**Comment by jamesknelson** [comment]
@JacksonGariety that sounds like an interesting solution. Could you explain in a little more detail what `navigation.pause()` would be supposed to do?

**Comment by OnkelTem** [comment]
I didn't get how to not leave the route. I want to make any form to be able to cancel any route change request.
So I put for example `{ unsavedData: true }` into context and.. then what? How to cancel route transition?

**Comment by OnkelTem** [comment]
> Nevermind. Redirecting back to the original component works.
> I had something in the component watching the route that caused it to re-render.

So you couldn't solve this task in the router, did you?

**Comment by OnkelTem** [comment]
I tried saving current route in the state but it didn't help too.

Say my current route is `/profiles/3/edit` which is a profile editing form.
When the form gets changed I set `app.unsavedData` flag in the state and store the current route:

```tsx
  "*": map(async (request, context: RouterContext) => {
    // Check for unsaved data
    const unsaved = context.store.getState().app.unsavedData;
    if (unsaved) {
      const currentRoute = context.store.getState().app.currentRoute; // == /profiles/3/edit
      return redirect(currentRoute);
    }
```

Now if I click say `/profiles/2` link, instead of following to the `/profiles/3/edit` according to the `redirect()` (i.e. staying where it was), it now returns 404 for the new route:

**NotFoundError: URL not found: /profiles/2**

`/profiles/2` actually does exist but for some reason it doesn't go there or stay where it was on `/profiles/3/edit`.

**Comment by OnkelTem** [comment]
@jamesknelson Do you have any idea? Navi is so powerful and asynchronous and flexible, I believe there should be a way to cancel a mere transition.

**Comment by OnkelTem** [comment]
ping @jamesknelson

**Comment by OnkelTem** [comment]
Dear James.

Would you please have a look at this thread and share your vision? In my humble opinion, this is very important part of the routing to be able to cancel transition.

**Comment by Menci** [comment]
ping @jamesknelson

**Comment by JacksonGariety** [comment]
Perhaps it should simply mimic the browser's `beforeunload` event. So navi should expose a method to set and unset a confirmation string, and then navi should display that confirmation string before navigating away from the route.

**Comment by Menci** [comment]
No, not just beforeunload. If you click "back" or "forward" button in
browser's navigation bar, NO EVENT could let you cancel the history pop.

**Comment by lukaszromerowicz** [comment]
Hi, is there any update on this? It still is an issue.

**Comment by mongris** [comment]
Thanks @lukaszromerowicz. I will try it later.
